src/main/java/com/google/cloud/genomics/dataflow/pipelines/AnnotateVariants.java

```diff
@@ -0,0 +1,307 @@
+  }
+
+  @Override
+  public void processElement(
+      DoFn<Contig, KV<String, VariantAnnotation>>.ProcessContext c) throws Exception {
+    Genomics genomics = auth.getGenomics(auth.getDefaultFactory());
+
+    Contig contig = c.element();
+    IntervalTree<Annotation> transcripts = retrieveTranscripts(genomics, contig);
+    ListMultimap<Range<Long>, Annotation> variantAnnotations =
+        retrieveVariantAnnotations(genomics, contig);
+    long baseOffset = transcripts.min().getStart();
+    String refBases = retrieveReferenceBases(genomics, refsetId,
+        contig.referenceName, baseOffset, transcripts.max().getEnd());
+
+    int varCount = 0;
+    Iterable<Variant> varIter = FluentIterable
+        .from(Paginator.Variants.create(genomics, ShardBoundary.STRICT).search(
+            c.element().getVariantsRequest(varsetId)
+              // TODO: Variants-only retrieval is not well support currently. For now
+              // we parameterize by CallSet for performance.
+              .setCallSetIds(callSetIds)
+              .setPageSize(VARIANTS_PAGE_SIZE),
+            VARIANT_FIELDS))
+        .filter(VariantUtils.IS_SNP);
+    for (Variant variant : varIter) {
+      List<String> alleles = ImmutableList.<String>builder()
+          .addAll(variant.getAlternateBases())
+          .add(variant.getReferenceBases())
+          .build();
+      Range<Long> pos = Range.openClosed(variant.getStart(), variant.getEnd());
+      for (String allele : alleles) {
+        String outKey = variant.getId() + ":" + allele;
+        for (Annotation match : variantAnnotations.get(pos)) {
+          if (allele.equals(match.getVariant().getAlternateBases())) {
+            // Exact match to a known variant annotation; straightforward join.
+            c.output(KV.of(outKey, match.getVariant()));
+          }
+        }
+        Iterator<Node<Annotation>> transcriptIter = transcripts.overlappers(
+            pos.lowerEndpoint().intValue(), pos.upperEndpoint().intValue() - 1); // Inclusive.
+        while (transcriptIter.hasNext()) {
+          // Calculate an effect of this allele on the coding region of the given transcript.
+          Annotation transcript = transcriptIter.next().getValue();
+          VariantEffect effect = AnnotationUtils.determineVariantTranscriptEffect(
+              variant.getStart(), allele, transcript, refBases.substring(
+                  (int) (transcript.getPosition().getStart() - baseOffset),
+                  (int) (transcript.getPosition().getEnd() - baseOffset)));
+          if (effect != null && !VariantEffect.SYNONYMOUS_SNP.equals(effect)) {
+            c.output(KV.of(outKey, new VariantAnnotation()
+                .setAlternateBases(allele)
+                .setType("SNP")
+                .setEffect(effect.toString())
+                .setGeneId(transcript.getTranscript().getGeneId())
+                .setTranscriptIds(ImmutableList.of(transcript.getId()))));
```

Comment:
This is just the very beginning, and there are a lot more opportunities to expand on here.  For instance, multiple sites are now currently ignored in [line 128 of AnnotationUtils.java](https://github.com/googlegenomics/dataflow-java/blob/variant-annotation/src/main/java/com/google/cloud/genomics/dataflow/utils/AnnotationUtils.java#L128):

```
if (codonExonOffset < 0 || exonBases.length() <= codonExonOffset+3) {
  LOG.fine("variant codon spans multiple exons, this case is not yet handled");
  return null;
}
```

There are SIFT and PolyPhen prediction to take into account.  Another option is to filter by minor allele frequency (MAF).  Again many other options to take into account to be sure the search is not too stringent, and yet as complete as possible for the dataset(s) and analysis being pursued.  

What about filtering on annotated subsets of data you generated and you deem they should be excluded as collection (set), based on analysis you generated?  This is a huge area, which is an understatement.

Paul